### PR TITLE
triton-kernels: add vendor-neutral Triton skill

### LINKED ISSUE #302: [FEATURE] Triton Skill
I believe a Triton skill would be complementary to the [cuda-kernel skill](https://github.com/huggingface/kernels/tree/main/skills/cuda-kernels), which would allow for more vendor independent development.
I do not currently have the expertise to implement this myself, so I am submitting this as a feature request to express interest and gauge demand.

### PR DESCRIPTION
**Pull request by jaygala223**
## Related issue

Closes #302

## What does this PR do?

Adds a new `triton-kernels` skill to `kernel-builder/skills/`, providing vendor-neutral guidance for writing portable Triton kernels that run on NVIDIA and AMD GPUs without modification.

## Motivation

The existing skills are either raw CUDA (`cuda-kernels`) or backend-specific Triton (`rocm-kernels`, `xpu-kernels`). There's no shared base covering portable Triton patterns that the backend-specific skills can build on. This was identified in #302 and the `skills/cuda-kernels` link in the original issue is now dead since skills moved to `kernel-builder/skills/` in #456.

## Changes

- Add `kernel-builder/skills/triton-kernels/SKILL.md`: core DSL patterns (program_id, masked loads, 2D tiling, reductions, tl.dot), autotune guidance, numerics (fp32 accumulation, safe softmax, dtype handling), reference kernel implementations (softmax, matmul, rmsnorm), benchmarking patterns, get_kernel integration, transformers patching, and common pitfalls table
- Add `references/autotune-guide.md`: deep dive on configs, key parameter, num_warps/num_stages selection, when NOT to autotune, practical config sets
- Add `references/kernel-patterns.md`: dropout (in-SRAM mask), fused add+rmsnorm, SiLU, SwiGLU, group-major PID ordering, pointer sliding
- Add `references/benchmarking-guide.md`: GB/s vs TFLOPS, roofline model, do_bench usage, when to expect gains over PyTorch
- Add `scripts/benchmark_template.py`: reusable benchmark harness with correctness check
- Add `scripts/correctness_template.py`: multi-shape, multi-dtype test suite template
- Add `examples/fused_softmax.py`: complete working kernel with test + benchmark
- Add `manifest.txt` and `CHANGELOG.md`

## Testing

- The example kernel (`examples/fused_softmax.py`) passes correctness tests against `torch.softmax` on V100 with irregular shapes (1823x781) at atol=1e-3
- Benchmark template runs and produces GB/s measurements
- Verified the skill follows the same directory structure as `rocm-kernels` (SKILL.md, manifest.txt, CHANGELOG.md, references/, scripts/, examples/)

## Checklist

- [x] This PR is linked to an issue that was discussed and approved
- [x] I have tested these changes locally
- [x] New/changed functionality has test coverage
- LLM disclosure:
  - [ ] I did not use an LLM to create this PR.
  - [x] I used an LLM for assistance while creating this PR.
  - [ ] This PR was mostly or completely generated by an LLM.

### CODE REVIEW

**Comment by sayakpaul** [comment]
Can you provide an example kernel that you wrote with this skill and tested? Feel free to apply to be able to publish `kernel` repo types on the Hub.

**Comment by jaygala223** [comment]
Hi @sayakpaul I tested the skill by using it as context for an AI agent and asking it to write a LayerNorm kernel (which is not one of the reference kernels already in the skill). The agent followed the patterns from the SKILL.md: row-wise reduction, fp32 accumulation, masked loads, contiguity assert, num_warps heuristic.

Results on V100 (fp32, M=4096):

Correctness: All shapes pass against `torch.nn.functional.layer_norm` at atol=1e-3, including irregular shapes (1823x781), single element, and bf16.

Performance: ~1.45x faster than PyTorch LayerNorm across all tested N values:

       N    Triton    PyTorch
     256    497       365 GB/s
    1024    727       548 GB/s
    2048    771       534 GB/s

The generated kernel + test is here: examples/skill_test_layernorm.py (https://github.com/huggingface/kernels/commit/8d7da2e0aad33acfd7a2862dc89b594af5f4374a)

I also applied for kernel repo access on the Hub.

**Comment by sayakpaul** [comment]
What is your Hub username?

**Comment by jaygala223** [comment]
jaygala223

**Comment by sayakpaul** [comment]
You have access to publish the kernels as proper kernel repos on the Hub. Could we please see the kernels generated with this skill on the Hub?

**Comment by jaygala223** [comment]
Yes. Published the layernorm kernel generated using the skill as a kernel repo: https://huggingface.co/kernels/jaygala223/triton-layernorm

**Comment by sayakpaul** [comment]
Are the results in https://github.com/huggingface/kernels/pull/675#issuecomment-4786468038 for https://huggingface.co/kernels/jaygala223/triton-layernorm?

**Comment by jaygala223** [comment]
Yes they are from the same kernel

**Review by sayakpaul — COMMENTED** [review]
We should also add skill to our docs.

**Comment by jaygala223** [comment]
Hi @sayakpaul added `triton-kernels` to the supported skills list in `docs/source/cli-skills.md`.

**Review by sayakpaul — COMMENTED** [review]
There should be a couple of changes skin to https://github.com/huggingface/kernels/pull/614.

**Comment by jaygala223** [comment]
Okay done!

Added a tip and example kernel.

File: docs/source/cli-skills.md [hunk 1]
```diff
@@ -6,6 +6,7 @@ Supported skills include:
 - `rocm-kernels`
 - `xpu-kernels`
 - `cpu-kernels`
+- `triton-kernels`
```
Comment (sayakpaul) [inline docs/source/cli-skills.md:9]: We're missing updates to 

* docs/source/builder/agents-guide.md
* docs/source/builder-cli.md
* kernel-builder/src/skills.rs

I already shared a reference PR: https://github.com/huggingface/kernels/pull/614

**Review by sayakpaul — CHANGES_REQUESTED** [review]

File: docs/source/cli-skills.md (hunk 1 above)
Comment (jaygala223) [reply docs/source/cli-skills.md:9]: Sorry. I missed that. Added it now. Thanks!

**Review by sayakpaul — APPROVED** [review]

**Comment by sayakpaul** [comment]
Failing tests are unrelated.
